Plugin for Dakota (#47)
watts-dev/watts: doc/source/user/plugins.rst

files. For example::

    results_reader = serpentTools.ResultsReader(str(result.outputs[-1]))

Dakota Plugin

> ++++++++++++++++

Reviewer: ```suggestion
Dakota Plugin
+++++++++++++
```

.. code-block:: jinja

    real = {{ real }}
    work_directory named = {{ workdir }}

Note that the execution of the Dakota plugin is slightly different than the
execution of the other plugins. Dakota is essentially an optimization and
uncertainty quantification tool that needs to be coupled to other external tools
or software. To run Dakota with WATTS, in addition to the WATTS Python script
and the Dakota input file, other files are necessary, including the input file
of the software that will be coupled with Dakota, the 'Dakota driver' Python
script, and any file necessary to run the coupled software.

If the templated Dakota input file is `dakota_watts_opt.in`, then the Dakota
plugin can be instantiated with the following command line::

    dakota_plugin = watts.PluginDakota('dakota_watts_opt.in')

If the coupled software uses a text-based input file, users can also template
this file with the `extra_template_inputs` options::

    dakota_plugin = watts.PluginDakota(
        template_file='dakota_watts_opt.in',
        extra_template_inputs=['extra_template_file_name'])

Prior to running Dakota with WATTS, the path to the 'dakota.sh' shell script

> needs to be provided either by adding it as `DAKOTA_DIR` to the environment or

Reviewer: ```suggestion
needs to be provided either by setting the :envvar:`DAKOTA_DIR` environment variable to the directory containing `dakota.sh` or
```

by adding it through the input file as::

    dakota_plugin.dakota_exec = "path/to/dakota.sh"

When the Dakota plugin is executed, it runs Dakota, which in turn executes the
'Dakota_driver' python script that facilitates the communication and data
transfer between Dakota and the coupled software. 'Dakota_driver' is a
simple Python script that requires only two inputs, namely the path to Dakota's
`interfacing` library and the coupled software's input file name. As its name
suggests, the 'interfacing' library is used by WATTS to interact between Dakota
and the coupled software.

Once the execution is complete, WATTS saves the results from all iterations as
individual objects and the final results as a separate object known as `finaldata1`

> in the :class:`~watts.Parameters` class.

Reviewer 2: The description of the DakotaPlugin is fine - however, we need to mention also that we can run the Dakota plugin through WATTS on a WATTS model, which is quite neat - enabling various levels of WATTS. I am wondering however if we can have connection from both levels or not (shared params).

files. For example::

    results_reader = serpentTools.ResultsReader(str(result.outputs[-1]))

Dakota Plugin
+++++++++++++

The :class:`~watts.PluginDakota` class handles execution of Dakota. Dakota uses
text-based input files that can be templated as follows:

.. code-block:: jinja

    real = {{ real }}
    work_directory named = {{ workdir }}

Note that the execution of the Dakota plugin is slightly different and involves

> more steps than the execution of the other plugins. Dakota is essentially an

Reviewer: ```suggestion
more steps than the execution of the other plugins. Dakota is an
```

optimization and uncertainty quantification tool that needs to be coupled to
other external tools or software.

> The execution of Dakota with WATTS is essentially a two-step process. In the

Reviewer: Sorry, too many "essentially"s :smile: 
```suggestion
The execution of Dakota with WATTS is a two-step process. In the
```

        extra_template_inputs=['extra_template_file_name', 'other_necessary_files'])

During the execution of WATTS, the working directory is switched to a temporary
location. Non-templated files needed by the coupled code (license file, data file,
etc.) can be copied to the temporary location with the `extra_inputs` option::

    dakota_plugin = watts.PluginDakota(
        template_file='dakota_watts_opt.in',
        extra_template_inputs=['extra_template_file_name', 'other_necessary_files'],
        extra_inputs=['file_1', 'file_2'])

In the Dakota input file, users need to provide the names of required files to the
`link_files` or the `copy_files` options where these files will be copied by Dakota
to the working directory during each iteration. Users can choose to input the names of
these files manually or they can choose to have WATTS automatically include all
file names in the `extra_template_inputs` and `extra_inputs` options. To do so, simply use
the `auto_link_files` option::

    dakota_plugin = watts.PluginDakota(
        template_file='dakota_watts_opt.in',
        extra_template_inputs=['extra_template_file_name', 'other_necessary_files'],

>         auto_link_files='<string_name_for_files>',

Reviewer 2: that looks good! thanks for taking such approach.
